From 17b17790888300531066dc9989001ee49d348c7e Mon Sep 17 00:00:00 2001
From: MengLi <meng.li@windriver.com>
Date: Mon, 29 Oct 2018 14:22:02 +0800
Subject: [PATCH] arch: arm: psci: implement psci_system_off interface for
 stratix10

This patch is missed when upgrade u-boot, so get it back from
WRlinux LTS 10.19.
When execute "poweroff" command in Linux, smc instruction is
invoked and then enter el3 mode to execute system off processing.
But in current code, there is only a weak define of psci_system_off,
and vendor doesn't implement this interface to power off system.
So, only make a workaround so that straxi10 board is able to run
command "poweroff" smoothly in linux.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-socfpga/psci.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/arch/arm/mach-socfpga/psci.c b/arch/arm/mach-socfpga/psci.c
index 10b09d3e7b..bd5f015b33 100644
--- a/arch/arm/mach-socfpga/psci.c
+++ b/arch/arm/mach-socfpga/psci.c
@@ -12,6 +12,8 @@
 #include <asm/arch/reset_manager_soc64.h>
 #include <asm/arch/rsu_s10.h>
 #include <asm/secure.h>
+#include <irq_func.h>
+#include <linux/delay.h>
 
 static u64 psci_cpu_on_64_cpuid __secure_data;
 static u64 psci_cpu_on_64_entry_point __secure_data;
@@ -163,3 +165,13 @@ int __secure psci_affinity_info_64(u32 function_id, u64 cpuid)
 {
 	return (psci_affinity_info_64_state >> (cpuid << 1)) & 0x03;
 }
+
+void __noreturn __secure psci_system_off(void)
+{
+	puts("poweroff ...\n");
+	udelay(50000); /* wait 50 ms */
+	disable_interrupts();
+
+	while (1)
+		;
+}
-- 
2.17.1

